File: 0014-longest-common-prefix/0014-longest-common-prefix.py
## divergence is the logic!
## Trie
class TrieNode:
    def __init__(self):
        self.child = dict()
        self.end = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def __insert__(self,word):
        current = self.root
        for w in word:
            if w not in current.child:
                current.child[w] = TrieNode()
            current = current.child[w]
        current.end = True
    
    def LCP(self):
        ans = ''
        current = self.root
        while True:
            if len(current.child)==1 and not current.end:
                key = next(iter(current.child))
                ans+=key
                current = current.child[key]
            else:
                break
        return ans
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        
            myTrie = Trie()
            for w in strs:
                myTrie.__insert__(w)
            
            return myTrie.LCP()
            
    
        

File: 0014-longest-common-prefix/0014-longest-common-prefix.java
class TrieNode{
    Map<Character,TrieNode> children;
    Boolean end;
    TrieNode(){
        children = new HashMap<>();
        end = false;
    }
}
class Trie{

    TrieNode root;
    Trie(){
        root = new TrieNode();
    }
    void insert(String s){
        TrieNode current = root;
        for(char c:s.toCharArray()){
            if (!current.children.containsKey(c)) {
                current.children.put(c, new TrieNode());
            }
            current = current.children.get(c);
        }
        current.end = true;
    }

    String LCP(){
        TrieNode curr = root;
        StringBuilder ans = new StringBuilder();
        while(true){
            if(curr.children.size()==1 && !curr.end){
                char key=curr.children.keySet().iterator().next();
                ans.append(key);
                curr = curr.children.get(key);
            }else{
                break;
            }
        }
        return ans.toString();
    }
    
}
class Solution {
    public String longestCommonPrefix(String[] strs) {
        Trie t =new Trie();
        for(String s:strs){
            t.insert(s);
        }
        return t.LCP();
    }
}